external-crates/move/move-compiler/src/cfgir/optimize/known_jumps.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    cfgir::{
+        ast::remap_labels,
+        cfg::{MutForwardCFG, CFG},
+    },
+    hlir::ast::{BasicBlocks, Command, Command_, FunctionSignature, Label, SingleType, Var},
+    shared::unique_map::UniqueMap,
+};
+
+use std::collections::{BTreeMap, BTreeSet};
+
+/// returns true if anything changed
+pub fn optimize(
+    _signature: &FunctionSignature,
+    _locals: &UniqueMap<Var, SingleType>,
+    cfg: &mut MutForwardCFG,
+) -> bool {
+    let changed = optimize_(cfg.blocks_mut());
+    if changed {
+        cfg.recompute();
+        remap_in_order(cfg.start_block(), cfg.blocks_mut());
+    }
+    changed
+}
+
+fn optimize_(blocks: &mut BasicBlocks) -> bool {
+    let known_jumps = find_known_jumps(blocks);
+    optimize_known_jumps(blocks, known_jumps)
+}
+
+fn find_known_jumps(blocks: &BasicBlocks) -> BTreeMap<Label, Label> {
+    use Command_ as C;
+    let mut forwarded_jumps = BTreeMap::new();
+    for (label, block) in blocks {
+        if block.len() == 1 {
+            match block.iter().last() {
+                Some(
+                    sp!(
+                        _,
+                        C::Jump {
+                            target,
+                            from_user: _
+                        }
+                    ),
+                ) => {
+                    forwarded_jumps.insert(*label, *target);
+                }
+                _ => (),
+            }
+        }
+    }
+
+    let mut known_jumps: BTreeMap<Label, Label> = BTreeMap::new();
+
+    for (start, target) in &forwarded_jumps {
+        let mut cur_target = *target;
+        let mut seen = BTreeSet::from([*target]);
+        while !known_jumps.contains_key(start) {
+            if known_jumps.contains_key(&cur_target) {
+                known_jumps.insert(*start, known_jumps[&cur_target]);
+            } else if start == &cur_target {
+                known_jumps.insert(*start, *start);
+            } else if !forwarded_jumps.contains_key(&cur_target) {
+                known_jumps.insert(*start, cur_target);
+            } else {
+                cur_target = forwarded_jumps[&cur_target];
+                if seen.contains(&cur_target) {
+                    // in a cycle, so bail
+                    known_jumps.insert(*start, cur_target);
+                } else {
+                    seen.insert(cur_target.clone());
+                }
+            }
+        }
```

Comment:
I found this a bit strange, but fine?
Just for consideration, I tried to change it a bit but my attempt at making it less strange?
```
for (start, target) in &forwarded_jumps {
    let mut cur_target = *target;
    let mut seen = BTreeSet::from([*target]);
    let final_target = loop {
        if known_jumps.contains_key(&cur_target) {
            break known_jumps[&cur_target];
        } else if start == &cur_target {
            break *start;
        } else if !forwarded_jumps.contains_key(&cur_target) {
            break cur_target;
        } else {
            let next = forwarded_jumps[&cur_target];
            if seen.contains(&next) {
                break next;
            } else {
                seen.insert(next.clone());
                cur_target = next;
            }
        }
    };
    known_jumps.insert(*start, final_target);
}
```

Reply:
Or 
```
for (start, target) in &forwarded_jumps {
    let mut cur_target = *target;
    let mut seen = BTreeSet::from([*target]);
    let final_target;
    loop {
        break if known_jumps.contains_key(&cur_target) {
            known_jumps[&cur_target]
        } else if start == &cur_target {
            *start
        } else if !forwarded_jumps.contains_key(&cur_target) {
            cur_target
        } else {
            let next = forwarded_jumps[&cur_target];
            if seen.contains(&next) {
                next
            } else {
                seen.insert(next.clone());
                cur_target = next;
                continue
            }
        }
    };
    known_jumps.insert(*start, final_target);
}
```